### PR TITLE
Implement read command

### PR DESCRIPTION
**Pull request by shaleh**
Resurrects https://github.com/helix-editor/helix/pull/3966/. This is 100% @idursun's work. Using `read_to_string` from https://github.com/helix-editor/helix/pull/7431.

closes https://github.com/helix-editor/helix/issues/3796

### CODE REVIEW

**Comment by the-mikedavis** [comment]
It's customary to give the original author Co-authored-by attribution when reusing work: https://docs.github.com/en/pull-requests/committing-changes-to-your-project/creating-and-editing-commits/creating-a-commit-with-multiple-authors

**Comment by shaleh** [comment]
I can do that. No problem.

File: helix-term/src/commands/typed.rs [hunk 1]
```diff
@@ -2454,6 +2455,37 @@ fn yank_diagnostic(
     Ok(())
 }
 
+fn read_file_info_buffer(
```
Comment (the-mikedavis) [inline helix-term/src/commands/typed.rs:2458]: There's a typo here info->into. But I think we can just call this `read`, it should be a little easier to find given the command name is `:read`

File: helix-term/src/commands/typed.rs [hunk 2]
```diff
@@ -2454,6 +2455,37 @@ fn yank_diagnostic(
...
+    args: &[Cow<str>],
+    event: PromptEvent,
+) -> anyhow::Result<()> {
+    if event != PromptEvent::Validate {
+        return Ok(());
+    }
+
+    let (view, doc) = current!(cx.editor);
+
+    ensure!(!args.is_empty(), "file name is expected");
```
Comment (the-mikedavis) [inline helix-term/src/commands/typed.rs:2469]: We can check `args.len() == 1` here instead so that we fail when there are multiple arguments

File: helix-term/src/commands/typed.rs [hunk 3]
```diff
@@ -2454,6 +2455,37 @@ fn yank_diagnostic(
...
+
+    let filename = args.get(0).unwrap();
+    let path = PathBuf::from(filename.to_string());
+    if !path.exists() {
+        bail!("file doesn't exist: {}", filename);
+    }
```
Comment (the-mikedavis) [inline helix-term/src/commands/typed.rs:2475]: Here we should also check whether the file is not a directory:

```suggestion
    ensure!(
        path.exists() && path.is_file(),
        "path is not a file: {:?}",
        path
    );
```

File: helix-term/src/commands/typed.rs [hunk 4]
```diff
@@ -2454,6 +2455,37 @@ fn yank_diagnostic(
...
+
+    let file = std::fs::File::open(path).map_err(|err| anyhow!("error reading file {}", err))?;
```
Comment (the-mikedavis) [inline helix-term/src/commands/typed.rs:2477]:
```suggestion
    let file = std::fs::File::open(path).map_err(|err| anyhow!("error opening file: {}", err))?;
```

File: helix-term/src/commands/typed.rs [hunk 5]
```diff
@@ -2454,6 +2455,37 @@ fn yank_diagnostic(
...
+    let mut reader = BufReader::new(file);
+    let (contents, _, _) = read_to_string(&mut reader, Some(doc.encoding()))
+        .map_err(|err| anyhow!("error reading file: {}", err))?;
+    let contents = Tendril::from(contents);
+    let selection = doc.selection(view.id);
+    let transaction = Transaction::insert(doc.text(), selection, contents);
+    doc.apply(&transaction, view.id);
```
Comment (the-mikedavis) [inline helix-term/src/commands/typed.rs:2484]:
```suggestion
    doc.apply(&transaction, view.id);
    doc.append_changes_to_history(view);
    view.ensure_cursor_in_view(doc, scrolloff);
```

Typable commands don't automatically do these steps so we need to do them explicitly. `scrolloff` is `cx.editor.config().scrolloff`

File: helix-term/tests/test/commands.rs [hunk 6]
```diff
@@ -640,3 +640,28 @@ async fn test_join_selections_space() -> anyhow::Result<()> {
...
+    let mut file = tempfile::NamedTempFile::new()?;
+    let contents_to_read = String::from("some contents");
+    let output_file = helpers::temp_file_with_contents(&contents_to_read)?;
+    let command = format!(":r {:?}<ret><esc>:w<ret>", output_file.path());
+
+    test_key_sequence(
+        &mut helpers::AppBuilder::new()
+            .with_file(file.path(), None)
+            .build()?,
+        Some(&command),
```
Comment (the-mikedavis) [inline helix-term/tests/test/commands.rs:655]: The command can be in line here:

```suggestion
        Some(&format!(":r {:?}<ret><esc>:w<ret>", output_file.path())),
```

I think it reads a little easier to have the command in the `test_key_sequence` helper

**Comment by the-mikedavis** [comment]
For the coauthor part, git/github is expecting the attribution in a `Name <email>` format so it should be:

```
Co-authored-by: Ibrahim Dursun <ibrahim@dursun.cc>
```

To fix the docs CI you can run `cargo xtask docgen` and commit the results

File: helix-term/src/commands/typed.rs (hunk 1 above)
Comment (shaleh) [inline helix-term/src/commands/typed.rs:2458]:
```suggestion
fn read(
```

File: helix-term/src/commands/typed.rs [hunk 7]
```diff
@@ -2454,6 +2455,37 @@ fn yank_diagnostic(
...
+    event: PromptEvent,
+) -> anyhow::Result<()> {
+    if event != PromptEvent::Validate {
+        return Ok(());
+    }
+
+    let (view, doc) = current!(cx.editor);
+
+    ensure!(!args.is_empty(), "file name is expected");
+
```
Comment (shaleh) [inline helix-term/src/commands/typed.rs:2470]:
```suggestion
ensure!(args.len() == 1, "only the file name is expected");

```

File: helix-term/src/commands/typed.rs [hunk 8]
```diff
@@ -3068,6 +3105,13 @@ pub const TYPABLE_COMMAND_LIST: &[TypableCommand] = &[
         fun: yank_diagnostic,
         signature: CommandSignature::all(completers::register),
     },
+    TypableCommand {
+        name: "read",
+        aliases: &["r"],
+        doc: "Load a file into buffer",
+        fun: read_file_info_buffer,
```
Comment (shaleh) [inline helix-term/src/commands/typed.rs:3112]:
```suggestion
        fun: read,
```

**Comment by shaleh** [comment]
Thank you for taking a look @the-mikedavis. All of the items identified have been resolved.

File: book/src/keymap.md [hunk 9]
```diff
@@ -24,6 +24,8 @@
 
 > 💡 Mappings marked (**TS**) require a tree-sitter grammar for the file type.
 
+> ⚠️ Some terminals' default key mappings conflict with Helix's. If any of the mappings described on this page do not work as expected, check your terminal's mappings to ensure they do not conflict. See the (wiki)[https://github.com/helix-editor/helix/wiki/Terminal-Support] for known conflicts.
```
Comment (pascalkuthe) [inline book/src/keymap.md:27]: some rebase went wrong because there are a bunch of unrelated commits and changes in this PR now

Comment (shaleh) [reply book/src/keymap.md:27]: yeah, I made the bonehead mistake of merging and pushing before leaving for the day. Murphy always takes his due. Sorry

File: helix-term/tests/test/commands.rs [hunk 10]
```diff
@@ -640,3 +640,27 @@ async fn test_join_selections_space() -> anyhow::Result<()> {
...
+        Some(&format!(":r {:?}<ret><esc>:w<ret>", output_file.path())),
+        Some(&|app| {
+            assert!(!app.editor.is_err(), "error: {:?}", app.editor.get_status());
+        }),
+        false,
+    )
+    .await?;
+
+    let expected_contents = format!("{}\n", contents_to_read);
+    helpers::assert_file_has_content(&mut file, &expected_contents)?;
```
Comment (the-mikedavis) [inline helix-term/tests/test/commands.rs:663]: This is failing on the windows CI since windows uses CRLF line endings. You can use the `LineFeedHandling` enum and `apply` to add the line ending here instead of `format!`

Comment (shaleh) [reply helix-term/tests/test/commands.rs]: pushed

File: helix-term/tests/test/commands.rs [hunk 11]
```diff
@@ -640,3 +640,27 @@ async fn test_join_selections_space() -> anyhow::Result<()> {
...
+        Some(&format!(":r {:?}<ret><esc>:w<ret>", output_file.path())),
+        Some(&|app| {
+            assert!(!app.editor.is_err(), "error: {:?}", app.editor.get_status());
+        }),
+        false,
+    )
+    .await?;
+
+    let expected_contents = LineFeedHandling::Native.apply(&contents_to_read);
+    helpers::assert_file_has_content(&mut file, &expected_contents)?;
```
Comment (the-mikedavis) [inline helix-term/tests/test/commands.rs:663]:
```suggestion
    let contents_to_read = "some contents";
    let output_file = helpers::temp_file_with_contents(contents_to_read)?;

    test_key_sequence(
        &mut helpers::AppBuilder::new()
            .with_file(file.path(), None)
            .build()?,
        Some(&format!(":r {:?}<ret><esc>:w<ret>", output_file.path())),
        Some(&|app| {
            assert!(!app.editor.is_err(), "error: {:?}", app.editor.get_status());
        }),
        false,
    )
    .await?;

    let expected_contents = LineFeedHandling::Native.apply(contents_to_read);
    helpers::assert_file_has_content(&mut file, &expected_contents)?;
```

`contents_to_read` is only used as a `&str` so we don't need to create a `String` from it

**Review by the-mikedavis — APPROVED** [review]

**Review by pascalkuthe — APPROVED** [review]
